### DISCUSSION TITLE
Automatic build and distribution system, pt. 2

### DISCUSSION THREAD

**Discussion by aried3r** [libvips/libvips#2011] [Ideas]
Hi!

I read https://github.com/jcupitt/libvips/issues/191 where [this PPA](https://launchpad.net/~dhor/+archive/ubuntu/myway) was referenced, however as of now, this is a bit outdated, the latest vips version there seems to be 8.5.7 released on 29 Jul 2017.

I was wondering if you would consider releasing your own packages via Travis CI. It seems you can deploy directly to launchpad when a new release is tagged. I'd love to help out, but I worry that most of the work would be on you with launchpad. :|

Resources:
https://docs.travis-ci.com/user/deployment/launchpad/
https://docs.travis-ci.com/user/deployment#Conditional-Releases-with-on%3A

I also [found some projects](https://github.com/search?q=%22provider%3A+launchpad%22+extension%3Ayml+path%3A%2F&type=Code) that are apparently using that:
https://github.com/Cybolic/vineyard/blob/2c10b6525f185e0eb5a9278a2c78362978097231/.travis.yml
https://github.com/phw/peek/blob/0c255e0cc33367790274f13157fe6e44d9a62ec1/.travis.yml

**Comment by jcupitt** [comment]
Hello Anton, 

Good idea. I registered libvips on launchpad and imported the source:

https://launchpad.net/libvips

I'll try and make a token for Travis.

I guess we'll need to make a set of Debian packages. Do you have any experience with this? I've always left deb packaging to the maintainer there.

**Comment by jcupitt** [comment]
Thinking about it again, we need some more stuff:

- we should run the libvips test suite automatically as well --- this is part of pyvips, so we need to pull that it, plus obviously Python
- we need to build at least four .deb packages (minimal binary, dev headers, command-line stuff, binary with vips7 compat)
- git master should build as "nightly", branches like "8.6" need to be followed separately and make .debs tagged as 8.6.5 (or whatever).

Perhaps there should be a libvips-launchpad project downstream of libvips containing all the scripts for this, and which triggers on every commit, and which uploads new .debs for every tag?

**Comment by aried3r** [comment]
Hey!

> I've always left deb packaging to the maintainer there.

Do you happen to be in contact with him? Could we ping him on GitHub? I checked Debian [unstable packages](https://packages.debian.org/source/sid/vips), but it seems there's also no current builds, we could otherwise use.

> Perhaps there should be a libvips-launchpad project downstream of libvips containing all the scripts for this, and which triggers on every commit, and which uploads new .debs for every tag?

You mean a new GitHub repo that gets triggered on new commits in this repo? If so, I'm unfamiliar with how this would work. I guess with a custom bash script that runs curl and triggers a build in that repo, see [this guide](https://docs.travis-ci.com/user/triggering-builds) but maybe there's a better way?

Btw, a friend just showed me [this](https://salsa.debian.org/kapouer/vips). It's being discussed in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=874759. Maybe there'll be some movement soon.

**Comment by jcupitt** [comment]
Update: the libvips test suite is now part of libvips, not pyvips, and is run on every commit. So that should make this issue a little simpler.

**Comment by aried3r** [comment]
Awesome!

There's also news on the debian side of things: https://tracker.debian.org/news/975313/accepted-vips-864-1-source-amd64-all-into-unstable/

**Comment by aried3r** [comment]
Ok, with the help of a friend I was able to build the necessary files for a PPA. I didn't test the last part, signing and uploading to Launchpad, since I don't have an account there yet.

So far I got it to build on Ubuntu 16.04 and 18.04 but not 14.04 which is what Travis offers. I guess we'd have to use a docker container and build it in there, if we want to automate it.

A few caveats:
- Haven't yet found the non-interactive version of `dch` which creates a changelog entry.
- Doesn't work on Ubuntu 14.04 (`mk-build-deps` finishes but doesn't actually do anything)
- Current "guide" (see below) uses Vagrant, not Docker and doesn't provide a `Dockerfile`

In case getting it to build automatically via Travis on git tags is too far off I was wondering what your current release process looks like. I saw that in the [releases](https://github.com/libvips/libvips/releases) you already attach some binaries, at least for Windows. Is that somehow automated, like having a `.build-and-release.sh`? Because if there is, maybe we (you, really :\) can also have an Ubuntu package released alongside it on Launchpad?

# Build instructions

## 18.04 Bionic

### On the host machine

```console
$ vagrant init ubuntu/bionic64
$ vagrant up
$ vagrant ssh
```

### In the Ubuntu VM

```bash
$ sudo apt-get update

# Add the gpg key of the vips package maintainer
$ sudo apt-get install debian-keyring
$ gpg --keyserver pool.sks-keyservers.net --recv-keys 7D887DC8BA7BBBA7B835E3BADCE310E7864CC8BF

# Get the necessary packages to build debian packages
$ sudo apt install devscripts

# Download the sources from Debian
$ dget http://deb.debian.org/debian/pool/main/v/vips/vips_8.7.0-1.dsc
# dget should also unpack to ./vips-8.7.0, but if it doesn't run
# dpkg-source -x vips_8.7.0-1.dsc

$ cd vips-8.7.0

# Creates a changelog entry for us
$ dch --local ~bionic "Rebuild for bionic."; dch -r -Dbionic

# Required as a dependency for mk-build-deps
$ sudo apt-get install equivs
# Install dependencies for libvips
$ mk-build-deps debian/control --install --root-cmd sudo --remove

# Build the package, will be created one directory above
$ dpkg-buildpackage -uc -us -S -sa

$ cd ..

# Sign the package
$ debsign vips_8.7.0-1~bionic1_source.changes

# Upload to PPA
$ dput ppa:<account>/<PPA-name> vips_8.7.0-1~bionic1_source.changes
```


## 16.04 Xenial

### On the host machine

```console
$ vagrant init ubuntu/xenial64
$ vagrant up
$ vagrant ssh
```

### In the Ubuntu VM

```bash
$ sudo apt-get update

# Add the gpg key of the vips package maintainer
$ sudo apt-get install debian-keyring
$ gpg --keyserver pool.sks-keyservers.net --recv-keys 7D887DC8BA7BBBA7B835E3BADCE310E7864CC8BF

# Get the necessary packages to build debian packages
$ sudo apt install devscripts

# On 16.04 we also need the newest version of debhelper
$ sudo apt-get install debhelper -t xenial-backports
# We also need libpng-dev, but that's not available in 16.04, but there's
# libpng12-dev, which was renamed to libpng-dev at some point.
# So we do a very dirty "hack" to meet the dependency on 16.04.
# But ideally we'd use equivs to provide a local package that provides
# the dependency. See https://unix.stackexchange.com/a/86251/125294
$ sed -i 's/libpng-dev/libpng12-dev/' debian/control

# Download the sources from Debian
$ dget http://deb.debian.org/debian/pool/main/v/vips/vips_8.7.0-1.dsc
# dget should also unpack to ./vips-8.7.0, but if it doesn't run
# dpkg-source -x vips_8.7.0-1.dsc

$ cd vips-8.7.0

# Creates a changelog entry for us
$ dch --local ~xenial "Rebuild for xenial."; dch -r -Dxenial

# Required as a dependency for mk-build-deps
$ sudo apt-get install equivs
# Install dependencies for libvips
$ mk-build-deps debian/control --install --root-cmd sudo --remove

# Build the package, will be created one directory above
$ dpkg-buildpackage -uc -us -S -sa

$ cd ..

# Sign the package
$ debsign vips_8.7.0-1~xenial1_source.changes

# Upload to PPA
$ dput ppa:<account>/<PPA-name> vips_8.7.0-1~xenial1_source.changes
```

**Comment by jcupitt** [comment]
Wow, fantastic! I'll try to look at this over the next couple of days.

**Comment by aried3r** [comment]
Small update, Travis now supports Ubuntu 16.04 https://changelog.travis-ci.com/ubuntu-xenial-16-04-build-environment-is-here!-79690 which should make building a lot easier. :)

**Comment by unikitty37** [comment]
Is there any update on this, please?

I'm trying to put a manual build up on launchpad.net just so I can get a recent libvips into a ruby-stretch Docker image, but the build instructions above fail for me at the `dput` stage:

```
root@8176df5ade5c:/usr/src/libvips# USER=unikitty37 dput ppa:unikitty37/libvips vips_8.7.2-1~bionic1_source.changes
Checking signature on .changes
gpg: /usr/src/libvips/vips_8.7.2-1~bionic1_source.changes: error 58: gpgme_op_verify
gpgme_op_verify: GPGME: No data
```

I'm completely new to building PPAs — could anybody please drop a hint on how to get this working? Thanks :)

**Comment by jcupitt** [comment]
If you're OK building in docker, it's easy to make a libvips from source:

https://github.com/jcupitt/docker-builds/blob/master/ruby-vips-stretch/Dockerfile

Of course a PPA would be better. 

I might have a stab at this issue in the next week or so.

**Comment by unikitty37** [comment]
Thanks — I do have my Dockerfile building from source at the moment. Unfortunately, while it runs fine in dev and staging, it fails under GitLab CI with this error:

```
      Failure/Error: require 'image_processing/vips'

      LoadError:
        Could not open library 'glib-2.0': /usr/lib/x86_64-linux-gnu/glib-2.0: cannot read file data: Is a directory.
        Could not open library 'libglib-2.0.so': libglib-2.0.so: cannot open shared object file: No such file or directory
      # ./cache/ruby/2.4.0/gems/ffi-1.9.25/lib/ffi/library.rb:145:in `block in ffi_lib'
      # ./cache/ruby/2.4.0/gems/ffi-1.9.25/lib/ffi/library.rb:99:in `map'
      # ./cache/ruby/2.4.0/gems/ffi-1.9.25/lib/ffi/library.rb:99:in `ffi_lib'
      # ./cache/ruby/2.4.0/gems/ruby-vips-2.0.13/lib/vips.rb:28:in `<module:GLib>'
      # ./cache/ruby/2.4.0/gems/ruby-vips-2.0.13/lib/vips.rb:13:in `<top (required)>'
```

This appears to be down to the way the ffi gem loads libraries, but I'm not sure if it's a missing dependency that's causing it — which is why I was hoping for a PPA, so it would resolve them on install :)

(I've disabled the libvips-related tests in CI for now so that builds can continue, but that's obviously not a good long-term solution!)
